src/coding-guidelines/types-and-traits.rst

```diff
@@ -63,3 +64,225 @@ Types and Traits
              current.checked_add(velocity).expect("Position calculation overflowed")
          }
 
+.. guideline:: Do Not Depend on Function Pointer Identity Across Crates
+    :id: gui_QbvIknd9qNF6 
+    :category: required
+    :status: draft
+    :release: unclear-latest
+    :fls: fls_1kg1mknf4yx7
+    :decidability: decidable
+    :scope: system
+    :tags: surprising-behavior
+
+    Do not rely on the equality or stable identity of function pointers originating from different crates or that may be inlined,
+    duplicated, or instantiated differently across compilation units, codegen units, or optimization profiles.
+
+    Avoid assumptions about low-level metadata (such as symbol addresses) unless explicitly guaranteed by the Ferrocene Language Specification (FLS).
+    Function address identity is not guaranteed by Rust and must not be treated as stable.
+    Rust’s ``fn`` type is a zero-sized function item promoted to a function pointer, whose address is determined by the compiler backend.
+    When a function resides in a different crate, or when optimizations such as inlining,
+    link-time optimization, or codegen-unit partitioning are enabled,
+    the compiler may generate multiple distinct code instances for the same function or alter the address at which it is emitted.
+
+    Consequently, the following operations are not reliable:
+
+    - Comparing function pointers for equality (``fn1 == fn2``)
+    - Assuming a unique function address
+    - Using function pointers as identity keys (e.g., in maps, registries, matchers)
+    - Matching behavior based on function address
+
+    This rule applies even when the functions are semantically identical, exported as ``pub``, or defined once in source form.
+
+    .. rationale:: 
+        :id: rat_xcVE5Hfnbb2u 
+        :status: draft
+
+        Compiler optimizations may cause function pointers originating from different crates to lose stable identity.
+        Observed behaviors include:
+
+        - Cross-crate inlining producing multiple code instantiations
+        - Codegen-unit separation causing function emission in multiple units
+        - Incremental builds producing variant symbol addresses
+        - Link-time optimization merging or splitting functions unpredictably
+
+        This behavior has resulted in real-world issues,
+        such as the bug reported in `rust-lang/rust#117047 <https://github.com/rust-lang/rust/issues/117047>`_,
+        where function pointer comparisons unexpectedly failed due to cross-crate inlining.
```

Comment:
Note that in the phrase "cross crate inlining", the term "inline" does not refer to the optimization where a call is replaced by the body of the callee, it refers to a strategy of lowering compiler IR in the caller crate instead of linking to it in the callee.

The term "inline" is used very confusingly. So in this case I would just say

>         where function pointer comparisons unexpectedly failed because the function in question was instantiated multiple times.

Reply:
Perhaps "instantiated" is a confusing term to those with a primarily C background, but it's supposed to be intuitive to those with more of a C++ mindset 🙃

Reply:
I can grok instantiated just fine.  Anyway, the target audience is rust programmer.
Resolved.